### ISSUE TITLE
Switch to uWebSockets

### ISSUE THREAD

**Issue by rclai** [meteor/meteor-feature-requests#75]
Big names like Deepstream, SocketCluster, Socket.io, and Discord are using it.

https://github.com/uNetworking/uWebSockets

Just something that stood out to me on the readme:

> One million WebSockets require ~111mb of user space memory (104 bytes per WebSocket).

Given Meteor's huge architectural dependence on WebSockets, it might make sense to explore it at least on a separate branch. I mean come on, you guys are upgrading to Node 8.

It's supposed to be a drop-in replacement to `ws`, so I suppose it might be a drop-in replacement for `faye-websocket` too.

**Comment by StorytellerCZ** [comment]
Would this addresses #41?

**Comment by rclai** [comment]
No, this is not to replace SockJS, it's to replace `faye-websocket` that Meteor uses.

**Comment by mitar** [comment]
But is it really the place to optimize? I mean, Meteor creates much more state in so many other places (subscriptions, merge box, etc.). Is it really a difference if we get websockets state down?

I think it would be useful to really measure how much large are different parts of server state per each client and then try to optimize based on that info.

**Comment by abernix** [comment]
While not a bad idea to consider alternatives, I'm fairly sure that `faye-websocket` is only used for server-to-server DDP connections in Meteor, which are far, far less common.  SockJS handles the bulk of the work in Meteor.

**Comment by rclai** [comment]
Oh whoops, I thought SockJS was just a client-side WebSocket polyfill for some reason.

**Comment by rclai** [comment]
@mitar, I didn't think about all that other stuff that Meteor does. I was only eyeing the performance/efficiency given from the low-level of µWebSockets.

**Comment by mitar** [comment]
Yes. Keeping state for each client is probably the main architectural difference to other approaches to web app development. As such it would be great if there would be more information about how much state is kept and ways to disable some of that. Like I am trying to [disable mergebox with this package](https://github.com/peerlibrary/meteor-control-mergebox).

**Comment by mitar** [comment]
> Oh whoops, I thought SockJS was just a client-side WebSocket polyfill for some reason

It is conceptually, but to make it work it has to provide also server side things (so that client can connect using polling for example).
